shell/src/main/java/alluxio/cli/bundler/CollectInfoAll.java

```diff
@@ -0,0 +1,300 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+/**
+ * Class for collecting various information about all nodes in the cluster.
+ */
+public class CollectInfoAll extends AbstractShell {
+  private static final Logger LOG = LoggerFactory.getLogger(CollectInfoAll.class);
+  private static final String FINAL_TARBALL_NAME =  "alluxio-cluster-info-%s.tar.gz";
+
+  private static final Map<String, String[]> CMD_ALIAS = ImmutableMap.of();
+
+  // In order for a warning to be displayed for an unstable alias, it must also exist within the
+  // CMD_ALIAS map.
+  private static final Set<String> UNSTABLE_ALIAS = ImmutableSet.of();
+
+  // TODO(jiacheng): what is a good max thread num?
+  private ExecutorService mExecutor;
+
+  /**
+   * Creates a new instance of {@link CollectInfoAll}.
+   *
+   * @param alluxioConf Alluxio configuration
+   */
+  public CollectInfoAll(InstancedConfiguration alluxioConf) {
+    super(CMD_ALIAS, UNSTABLE_ALIAS, alluxioConf);
+  }
+
+  // TODO(jiacheng): phase 2 specify hosts from cmdline
+  /**
+   * Finds all hosts in the Alluxio cluster.
+   * We assume the masters and workers cover all the nodes in the cluster.
+   *
+   * @return a set of hostnames in the cluster
+   * */
+  public Set<String> getHosts() {
+    String confDirPath = mConfiguration.get(PropertyKey.CONF_DIR);
+    System.out.format("Looking for masters and workers in %s%n", confDirPath);
+    Set<String> hosts = new HashSet<>();
+    hosts.addAll(CommandUtils.readNodeList(confDirPath, "masters"));
+    hosts.addAll(CommandUtils.readNodeList(confDirPath, "workers"));
+    System.out.format("Found %s hosts%n", hosts.size());
+    return hosts;
+  }
+
+  /**
+   * Main method, starts a new CollectInfoAll shell.
+   * CollectInfoAll will SSH to all hosts and invoke {@link CollectInfo}.
+   * Then collect the tarballs generated on each of the hosts to the localhost.
+   * And tarball all results into one final tarball.
+   *
+   * @param argv array of arguments given by the user's input from the terminal
+   */
+  public static void main(String[] argv) throws IOException {
+    int ret = 0;
+
+    InstancedConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+
+    // Execute the Collectors one by one
+    // Reduce the RPC retry max duration to fail earlier for CLIs
+    conf.set(PropertyKey.USER_RPC_RETRY_MAX_DURATION, "5s", Source.DEFAULT);
+    CollectInfoAll shellAll = new CollectInfoAll(conf);
+
+    // Validate commands
+    if (argv.length < 2) {
+      shellAll.printUsage();
+      System.exit(-1);
+    }
+    String targetDir = argv[1];
+
+    List<String> allHosts = new ArrayList<>(shellAll.getHosts());
+    System.out.format("Init thread pool for %s hosts%n", allHosts.size());
+    shellAll.mExecutor = Executors.newFixedThreadPool(allHosts.size());
```

Comment:
Is this a reasonable thread pool? What if the cluster is very big?

Reply:
probably okay for the execution stage where we are just waiting for commands to be executed on the remote servers. It might be too many concurrent connections trying to copy files back from large cluster.

Reply:
Due to resource limitation I can only test with a small cluster with probably big log files (maybe 100 log files rotation). Personally I prefer to optimize this afterwards, with real feedback from users. @ns1123 what do you think?

Reply:
Can we add an option to the command to limit the maximum thread pool size? Then here we can do something like `shellAll.mExecutor = Executors.newFixedThreadPool(Math.min(allHosts.size(), maxThreads))`